sqlglot/optimizer/scope.py

```diff
@@ -0,0 +1,274 @@
+        """Rename a selectable in this scope"""
+        columns = self.selectables.pop(old_name or "", [])
+        self.selectables[new_name] = columns
+
+
+def traverse_scope(expression):
+    """
+    Traverse an expression by it's "scopes".
+
+    "Scope" represents the current context of a Select statement.
+
+    This is helpful for optimizing queries, where we need more information than
+    the expression tree itself. For example, we might care about the selectable
+    names within a subquery.
+
+    Examples:
+        >>> import sqlglot
+        >>> expression = sqlglot.parse_one("SELECT a FROM (SELECT a FROM x) AS y")
+        >>> scopes = list(traverse_scope(expression))
+        >>> scopes[0].expression.sql(), list(scopes[0].selectables)
+        ('SELECT a FROM x', ['x'])
+        >>> scopes[1].expression.sql(), list(scopes[1].selectables)
+        ('SELECT a FROM (SELECT a FROM x) AS y', ['y'])
+
+    Args:
+        expression (exp.Expression): expression to traverse
+    Yields:
+        Scope: scope instances
+    """
+    yield from _traverse_scope(Scope(expression))
+
+
+def _traverse_scope(scope):
+    # Collect all the relevant Expression parts in this scope
+    for node, parent, _ in _bfs_until_next_scope(scope.expression):
+        if node is scope.expression:
+            continue  # Skip this node itself - we only care about children
+        if isinstance(node, (exp.Select, exp.Union)):
+            if isinstance(parent, exp.CTE):
+                scope.ctes.append(parent)
+            elif isinstance(parent, exp.Subquery):
+                scope.derived_tables.append(parent)
+            else:
+                scope.subqueries.append(node)
+        elif isinstance(node, exp.Table):
+            scope.tables.append(node)
+        elif isinstance(node, exp.Column) and not isinstance(node.this, exp.Star):
+            scope.columns.append(node)
+
+    if isinstance(scope.expression, exp.Select):
+        yield from _traverse_select(scope)
+    elif isinstance(scope.expression, exp.Union):
+        yield from _traverse_union(scope)
+    else:
+        raise OptimizeError(f"Unexpected expression type: {scope.expression}")
+    yield scope
+
+
+def _traverse_select(scope):
+    yield from _traverse_derived_tables(scope.ctes, scope)
+    yield from _traverse_derived_tables(scope.derived_tables, scope)
+
+    for table in scope.tables:
+        table_name = table.text("this")
+
+        if isinstance(table.parent, exp.Alias):
+            selectable_table = table.parent.alias
+        else:
+            selectable_table = table_name
+
+        if table_name in scope.selectables:
+            # This is a reference to a parent selectable (e.g. a CTE), not an actual table.
+            scope.selectables[selectable_table] = scope.selectables[table_name]
+        elif selectable_table in scope.selectables:
+            raise OptimizeError(f"Duplicate table name: {selectable_table}")
```

Comment:
is this actually an error?

Reply:
Again, I'm not too sure about standards here. Do you have a good handle on what we should do?